commit 5c2465dfd457f3015eebcc3ace50570e1d896aeb
Author: Chuck Lever <chuck.lever@oracle.com>
Date:   Thu Jul 15 15:52:12 2021 -0400

    SUNRPC: Set rq_auth_stat in the pg_authenticate() callout
    
    In a few moments, rq_auth_stat will need to be explicitly set to
    rpc_auth_ok before execution gets to the dispatcher.
    
    svc_authenticate() already sets it, but it often gets reset to
    rpc_autherr_badcred right after that call, even when authentication
    is successful. Let's ensure that the pg_authenticate callout and
    svc_set_client() set it properly in every case.
    
    Signed-off-by: Chuck Lever <chuck.lever@oracle.com>
    Signed-off-by: Anna Schumaker <Anna.Schumaker@Netapp.com>

diff --git a/fs/lockd/svc.c b/fs/lockd/svc.c
index 2de048f80eb8..8e936999216c 100644
--- a/fs/lockd/svc.c
+++ b/fs/lockd/svc.c
@@ -649,6 +649,7 @@ static int lockd_authenticate(struct svc_rqst *rqstp)
 	switch (rqstp->rq_authop->flavour) {
 		case RPC_AUTH_NULL:
 		case RPC_AUTH_UNIX:
+			rqstp->rq_auth_stat = rpc_auth_ok;
 			if (rqstp->rq_proc == 0)
 				return SVC_OK;
 			if (is_callback(rqstp->rq_proc)) {
@@ -659,6 +660,7 @@ static int lockd_authenticate(struct svc_rqst *rqstp)
 			}
 			return svc_set_client(rqstp);
 	}
+	rqstp->rq_auth_stat = rpc_autherr_badcred;
 	return SVC_DENIED;
 }